src/track/beatgrid.cpp

```diff
@@ -291,39 +282,27 @@ double BeatGrid::getBpmAroundPosition(double curSample, int n) const {
     Q_UNUSED(curSample);
     Q_UNUSED(n);
 
-    QMutexLocker locker(&m_mutex);
     if (!isValid()) {
         return -1;
     }
     return bpm();
 }
 
-void BeatGrid::addBeat(double dBeatSample) {
-    Q_UNUSED(dBeatSample);
-    //QMutexLocker locker(&m_mutex);
-    return;
-}
-
-void BeatGrid::removeBeat(double dBeatSample) {
-    Q_UNUSED(dBeatSample);
-    //QMutexLocker locker(&m_mutex);
-    return;
-}
-
-void BeatGrid::translate(double dNumSamples) {
-    QMutexLocker locker(&m_mutex);
+BeatsPointer BeatGrid::translate(double dNumSamples) const {
     if (!isValid()) {
-        return;
+        return BeatsPointer(new BeatGrid(*this));
```

Comment:
This clone is redundant, theoretically. But it is the only safe solution now.

https://en.cppreference.com/w/cpp/memory/enable_shared_from_this/shared_from_this

Unfortunately this operation is not safe and could cause undefined behavior.

Maybe use `immer::box` instead of `std::shared_ptr` for sharing immutable objects?

https://sinusoid.es/immer/containers.html#box

Reply:
We may add a static wrapper around this function to do the isValid() check.